### PR TITLE
[NMS] Refresh equipment and subscriber state every 30 sec

### LINKED ISSUE #3708: NMS UI not auto refreshing status of gateways
### Your Environment

- **Version:** v1.3
- **Affected Component:**  NMS
- **Affected Subcomponent:** page <tenant>/equipment/overview/gateway
- **Deployment Environment:** AWS (orc8r)

### Describe the Issue

The NMS page does not auto-refresh the status of Gateways. This causes "good/green" to switch to "bad/red" after ~5 min. If the user refreshes the browser page, the status refreshes and becomes correct again.  

**To Reproduce**
1. Go to <tenant>/equipment/overview/gateway in NMS
2. Note the checkin time / status for each gateway displayed
3. Let the page sit for 5min or more
4. See that status has changed to red/bad for any AGW that had been green and last checkin time is old
5. refresh the browser page, see that AGWs that were just "bad" are now "good" again. 

**Expected behavior**
The expected behavior is that page triggers browser pulls data periodically on its own to keep status accurate. 

I have tested with Chrome, Safari, and Firefox, so I don't think it is browser specific issue.

### PR DESCRIPTION
**Pull request by HannaFar**
## Summary

Refresh gateway context every 30 sec

### CODE REVIEW

**Review by andreilee — CHANGES_REQUESTED** [review]
Please add your testing procedure description to the pull request

File: nms/app/packages/magmalte/app/components/lte/LteContext.js [hunk 1]
```diff
@@ -96,6 +96,12 @@ export function GatewayContextProvider(props: Props) {
       setIsLoading(false);
     };
     fetchState();
+    // Refresh data every 30 seconds
+    const interval = setInterval(() => {
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/lte/LteContext.js]: Instead of hardcoding the time here, can you use a variable here, we can perhaps move this to a configurable settings component later on. 

Additionally I was wondering if we can trigger this more explicitly through a context consumer, rather than autorefreshing by default. Thanks

File: nms/app/packages/magmalte/app/components/context/GatewayContext.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ export type GatewayContextType = {
   state: {[string]: lte_gateway},
   setState: (key: gateway_id, val?: mutable_lte_gateway) => Promise<void>,
   updateGateway: (props: $Shape<UpdateGatewayProps>) => Promise<void>,
+  refreshSate?: (interval: number, id?: string) => Promise<void>,
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/GatewayContext.js]: minor nit: Fix typo - refreshSate -> refreshState

File: nms/app/packages/magmalte/app/views/equipment/EnodebDetailMain.js [hunk 3]
```diff
@@ -18,26 +18,28 @@ import type {WithAlert} from '@fbcnms/ui/components/Alert/withAlert';
...
 import Grid from '@material-ui/core/Grid';
 import React from 'react';
 import SettingsIcon from '@material-ui/icons/Settings';
-import SettingsInputAntennaIcon from '@material-ui/icons/SettingsInputAntenna';
+import Text from '../../theme/design-system/Text';
 import TopBar from '../../components/TopBar';
 import nullthrows from '@fbcnms/util/nullthrows';
 import withAlert from '@fbcnms/ui/components/Alert/withAlert';
 
+import moment from 'moment';
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/views/equipment/EnodebDetailMain.js]: minor nit: Move the extra line above to be between the 
`import moment from 'moment';` and `import {DateTimePicker} from '@material-ui/pickers';`
Similarly remove additional empty lines above and separate the type imports and rest of the imports with a empty line

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  //$FlowIgnore
+  refreshingState?: any,
+  enqueueSnackbar?: (msg: string, cfg: {}) => ?(string | number),
+};
+
+type refreshType = 'subscriber' | 'gateway' | 'enodeb';
+
+export function useRefreshingContext(props: Props) {
+  const ctx = useContext(props.context);
+  const [refreshingState, setRefreshingState] = useState(ctx);
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: Since we are already in useRefreshingContext hook, We can call this just state and setState

File: nms/app/packages/magmalte/app/components/lte/LteContext.js [hunk 5]
```diff
@@ -106,13 +107,14 @@ export function GatewayContextProvider(props: Props) {
...
-        setState: (key, value?) => {
+        setState: (key, value?, refreshingState?: {[string]: lte_gateway}) => {
           return SetGatewayState({
             lteGateways,
             setLteGateways,
             networkId,
-            key,
+            key: key,
             value,
+            refreshingState: refreshingState,
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/lte/LteContext.js]: Instead of calling this refreshing state, can you call this newState and you can just use the shorthand dictionary notation and call setGatewayState with (lteGateways, setLteGateways, networkId, key, value, newState)

File: nms/app/packages/magmalte/app/components/lte/LteContext.js [hunk 6]
```diff
@@ -161,8 +163,15 @@ export function EnodebContextProvider(props: Props) {
...
-        setState: (key, value?) =>
-          SetEnodebState({enbInfo, setEnbInfo, networkId, key, value}),
+        setState: (key: string, value?, refreshingState?: EnodebState) =>
+          SetEnodebState({
+            enbInfo,
+            setEnbInfo,
+            networkId,
+            key,
+            value,
+            refreshingState,
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/lte/LteContext.js]: same comment as above

File: nms/app/packages/magmalte/app/views/equipment/EnodebDetailSummaryStatus.js [hunk 7]
```diff
@@ -36,15 +42,30 @@ export function EnodebSummary() {
...
-  const enbInfo = ctx.state.enbInfo[enodebSerial];
+  const networkId: string = nullthrows(match.params.networkId);
+  const enqueueSnackbar = useEnqueueSnackbar();
 
+  // Auto refresh enodeb every 30 seconds
+  const ctx = useRefreshingContext({
+    context: EnodebContext,
+    networkId: networkId,
+    type: 'enodeb',
+    interval: 30000,
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/views/equipment/EnodebDetailSummaryStatus.js]: Can you set this to a constant instead of assigning it to a value directly.

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+      props.context.setState(
+        null,
+        props.networkId,
+        props.refreshingState.state,
+      );
+    }
+  }
+}
+
+type fetchProps = {
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: Can you move the fetch part of the code under app/state directory. Also this logic seems duplicated, we should merge the Init and the fetch methods.

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            ...refreshingState.state,
+            enbInfo: {
+              ...refreshingState.state.enbInfo,
+              [id]: {enb_state: newEnbSt ?? {}, enb: newEnb},
+            },
+          },
+        };
+      }
+    } catch (e) {
+      enqueueSnackbar?.('failed fetching enodeb informations', {
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: enodeb state not being present is a valid case, let's remove this snackbar error.

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+          },
+        },
+      };
+    }
+  } else {
+    let enb = {};
+    try {
+      enb = await MagmaV1API.getLteByNetworkIdEnodebs({networkId});
+    } catch (e) {
+      enqueueSnackbar?.('failed fetching enodeb information', {
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: let's ignore this error

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+        };
+      }
+    } catch (e) {
+      enqueueSnackbar?.('failed fetching gateway informations', {
+        variant: 'error',
+      });
+    }
+  }
+}
+async function fetchSubscribers(props: fetchProps) {
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js]: This might be missing the logic currently present in InitSubscriberState, As mentioned earlier, we should probably merge the init and the fetch methods. https://github.com/magma/magma/blob/master/nms/app/packages/magmalte/app/state/lte/SubscriberState.js#L35

**Review by karthiksubraveti — CHANGES_REQUESTED** [review]
Thanks for making this change. The changes mostly look good. Left a few comments.

File: nms/app/packages/magmalte/app/state/lte/EquipmentState.js [hunk 12]
```diff
@@ -108,58 +108,88 @@ export async function SetTierState(props: TierStateProps) {
...
-      );
-      return [enb[k], enbSt ?? {}];
-    } catch (e) {
-      return [enb[k], {}];
-    }
-  });
+type InitEnodeStateProps = {
+  networkId: network_id,
+  setEnbInfo: ({[string]: EnodebInfo}) => void,
+  enqueueSnackbar?: (msg: string, cfg: {}) => ?(string | number),
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/state/lte/EquipmentState.js:182]: minor nit: We aren't using enqueueSnackbar, we can get rid of this.

File: nms/app/packages/magmalte/app/state/lte/EquipmentState.js [hunk 13]
```diff
@@ -192,19 +222,6 @@ export async function SetEnodebState(props: EnodebStateProps) {
       const prevEnbSt = enbInfo[key].enb_state;
       setEnbInfo({...enbInfo, [key]: {enb_state: prevEnbSt, enb: value.enb}});
     }
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/state/lte/EquipmentState.js:224]: Why are we making this change ? is this because the state should get refreshed anyways ?

File: nms/app/packages/magmalte/app/state/lte/EquipmentState.js [hunk 14]
```diff
@@ -213,10 +230,44 @@ export async function SetEnodebState(props: EnodebStateProps) {
     const newEnbInfo = {...enbInfo};
     delete newEnbInfo[key];
     setEnbInfo(newEnbInfo);
+    return;
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/state/lte/EquipmentState.js:233]: unnecessary return. we can get rid fo this

File: nms/app/packages/magmalte/app/views/equipment/EnodebDetailMain.js [hunk 15]
```diff
@@ -160,45 +163,99 @@ function EnodebRebootButtonInternal(props: WithAlert) {
...
+  const enodebSerial: string = nullthrows(match.params.enodebSerial);
+  const enbInfo = ctx.state.enbInfo[enodebSerial];
+  const enbIpAddress = enbInfo?.enb_state?.ip_address ?? '';
+
+  return (
+    <DateTimeMetricChart
+      title={CHART_TITLE}
+      unit={'Throughput(mb/s)'}
+      queries={[
+        `rate(gtp_port_user_plane_dl_bytes{service="pipelined", ip_addr="${enbIpAddress}"}[5m])/1000`,
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/views/equipment/EnodebDetailMain.js:251]: Looks like the conflict here is incorrectly resolved. Can you add sum(rate(...)) here.

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    const newState = await fetchRefreshState({
+      context: props.context,
+      type: props.type,
+      networkId: props.networkId,
+      id: props.id,
+      enqueueSnackbar: props.enqueueSnackbar,
+    });
+    if (newState) {
+      setState(prevState => {
+        if (props.type === 'subscriber') {
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js:54]: minor nit: can you define a constant here.

File: nms/app/packages/magmalte/app/components/context/RefreshContext.js [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+  id?: string,
+  enqueueSnackbar?: (msg: string, cfg: {}) => ?(string | number),
+  refresh: boolean,
+  lastRefreshTime?: string,
+};
+
+type refreshType = 'subscriber' | 'gateway' | 'enodeb';
+
+export function useRefreshingContext(props: Props) {
+  const ctx = useContext(props.context);
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/components/context/RefreshContext.js:41]: Why did you change the original logic of refreshing the context at the end of the destructor ? What was the issue there ?

File: nms/app/packages/magmalte/app/views/subscriber/SubscriberOverview.js [hunk 18]
```diff
@@ -174,16 +189,83 @@ function JsonDialog(props: Props) {
...
+                  alignItems="center"
+                  spacing={2}>
+                  <Grid item>
+                    <AddSubscriberButton />
+                  </Grid>
+                  <Grid item>
+                    <FormControlLabel
+                      value="top"
+                      control={
+                        <SwitchButton
```
Comment (karthiksubraveti) [inline nms/app/packages/magmalte/app/views/subscriber/SubscriberOverview.js:224]: Can you move this before AddSubscriberButton ? Also can we use the Autorefresh checked button consistently here as well or modify that page to use the switch. I''m referring to the Gateway Event Table and GatewayLog Table

**Comment by karthiksubraveti** [comment]
These changes were discarded. Closing this one to avoid confusion
